bdsx/bds/player.ts

```diff
@@ -248,6 +248,122 @@ export class Player extends Actor {
     canDestroy(block: Block): boolean {
         abstract();
     }
+
+    /**
+     * Returns the player's XP points
+     */
+    getXp(): number {
+        return Math.round(this.getXpProgress() * this.getXpNeededForNextLevel());
+    }
+
+    /**
+     * Returns the player's progression to the next level, between 0.0 and 1.0
+     */
+    getXpProgress(): number {
+        return this.getAttribute(AttributeId.PlayerExperience);
+    }
+
+    /**
+     * Returns the player's XP level
+     */
+    getXpLevel(): number {
+        return this.getAttribute(AttributeId.PlayerLevel);
+    }
+
+    /**
+     * Sets the player's XP points
+     *
+     * @param xp - between 1 and the maximum XP points for the level
+     */
+    setXp(xp: number): void {
+        this.setAttribute(AttributeId.PlayerExperience, xp / this.getXpNeededForNextLevel() > 1 ? 1 : xp / this.getXpNeededForNextLevel());
+    }
+
+    /**
+     * Sets the player's progression to the next XP level
+     *
+     * @param progress - between 0.0 and 1.0
+     */
+    setXpProgress(progress: number): void {
+        this.setAttribute(AttributeId.PlayerExperience, progress > 1 ? 1 : progress);
+    }
+
+    /**
+     * Sets the player's XP level
+     *
+     * @param level - between 0 and 24791
+     */
+    setXpLevel(level: number): void {
+        this.setAttribute(AttributeId.PlayerLevel, level > 24791 ? 24791 : level < 0 ? 0 : level);
+    }
+
+    /**
+     * Adds XP points to the player, recalculating their level & progress
+     *
+     * @param xp - XP to add
+     */
+    addXp(xp: number): void {
+        abstract();
+    }
+
+    /**
+     * Adds progress to the player's XP level
+     *
+     * @param progress - between 0.0 and 1.0
+     */
+    addXpProgress(progress: number): void {
+        this.setAttribute(AttributeId.PlayerExperience, this.getXpProgress() + progress > 1 ? 1 : this.getXpProgress() + progress);
+    }
```

Comment:
```suggestion
    addXpProgress(progress: number): void {
        if (-progress > this.getXpProgress()) this.setXpProgress(0);
        else this.setAttribute(AttributeId.PlayerExperience, this.getXpProgress() + progress > 1 ? 1 : this.getXpProgress() + progress);
    }
```
Let's make this support negative numbers

Reply:
By consistency with other functions I meant the functions I added
By following the native functions names we would have for example `addExperience()` but `getXpNeededForNextLevel()`, so a mix of `Xp` and `Experience`. But yeah that's not really a problem. I'm fixing this